### ISSUE TITLE
haskell-ghc, haskell-stack-ghc fail to find autogenerated Paths_pkgname module

### ISSUE THREAD

**Issue by andersk** [flycheck/flycheck#1346]
For every Cabal package `pkgname`, Cabal automatically generates a [`Paths_pkgname` module](https://www.haskell.org/cabal/users-guide/developing-packages.html#accessing-data-files-from-package-code) that can be imported within that package. But the `haskell-ghc` and `haskell-stack-ghc` checkers show `error: Failed to load interface for ‘Paths_pkgname’` on such an import line (thus preventing it from finding any other errors or warnings), even though `cabal build` or `stack build` works fine.

There are plenty of examples of this in the wild, such as [hlint](https://github.com/ndmitchell/hlint/blob/v2.0.9/src/CmdLine.hs#L28). Here’s a minimal example for testing.

**`paths-test.cabal`**

```
name:                paths-test
version:             0.1.0.0
build-type:          Simple
cabal-version:       >= 1.10

executable paths-test
  main-is:             Main.hs
  build-depends:       base
  default-language:    Haskell2010
```

**`Main.hs`**

```haskell
import qualified Paths_paths_test as P
main = do
  print P.version
  putStrLn =<< P.getBinDir
  putStrLn =<< P.getLibDir
  putStrLn =<< P.getDynLibDir
  putStrLn =<< P.getDataDir
  putStrLn =<< P.getLibexecDir
  putStrLn =<< P.getSysconfDir
```

You can either `cabal build` or `stack init; stack build` this. (The autogenerated source is placed at `dist/build/autogen/Paths_paths_test.hs` or `.stack-work/dist/x86_64-linux-nopie/Cabal-1.24.2.0/build/autogen/Paths_paths_test.hs`, and compiled to `dist/build/paths-test/paths-test-tmp/Paths_paths_test.hi` or `.stack-work/dist/x86_64-linux-nopie/Cabal-1.24.2.0/build/paths-test/paths-test-tmp/Paths_paths_test.hi`.)

But Flycheck shows an error on the line that imports `Paths_paths_test`.

Output of `M-x flycheck-compile haskell-ghc`:

```
-*- mode: compilation; default-directory: "~/haskell/paths-test/" -*-
Compilation started at Wed Oct 18 04:30:22

ghc -Wall -no-link -outputdir /tmp/flycheck-haskell-ghc-cache2472-xw -i/home/anders/haskell/paths-test/ -x hs /home/anders/haskell/paths-test/Main.hs
[1 of 1] Compiling Main             ( /home/anders/haskell/paths-test/Main.hs, /tmp/flycheck-haskell-ghc-cache2472-xw/Main.o )

/home/anders/haskell/paths-test/Main.hs:1:1: error:
    Failed to load interface for ‘Paths_paths_test’
    Use -v to see a list of the files searched for.

Compilation exited abnormally with code 1 at Wed Oct 18 04:30:22
```

Output of `M-x flycheck-compile haskell-stack-ghc`:

```
-*- mode: compilation; default-directory: "~/haskell/paths-test/" -*-
Compilation started at Wed Oct 18 04:30:40

stack ghc -- -Wall -no-link -outputdir /tmp/flycheck-haskell-ghc-cache2472-xw -i/home/anders/haskell/paths-test/ -x hs /home/anders/haskell/paths-test/Main.hs
[1 of 1] Compiling Main             ( /home/anders/haskell/paths-test/Main.hs, /tmp/flycheck-haskell-ghc-cache2472-xw/Main.o )

/home/anders/haskell/paths-test/Main.hs:1:1: error:
    Failed to load interface for ‘Paths_paths_test’
    Use -v to see a list of the files searched for.

Compilation exited abnormally with code 1 at Wed Oct 18 04:30:40
```

Output of `M-x flycheck-verify-setup`:

```
Syntax checkers for buffer Main.hs in haskell-mode:

  haskell-stack-ghc
    - may enable: yes
    - executable: Found at /home/anders/.local/bin/stack

  haskell-ghc
    - may enable: yes
    - executable: Found at /usr/bin/ghc

  haskell-hlint
    - may enable:         yes
    - executable:         Found at /home/anders/.local/bin/hlint
    - configuration file: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20171009.1145)
Emacs version:    25.2.2
System:           x86_64-pc-linux-gnu
Window system:    x
```

I’m using emacs25 25.2+1-6 on Ubuntu 17.10, with flycheck 20171009.1145 from MELPA.

**Comment by andersk** [comment]
Apparently the flycheck-haskell extension fixes this, so maybe this can be closed.

**Comment by fmdkdd** [comment]
Thanks for the nicely detailed report.  The core `flycheck` package tries to avoid doing any involved detection of project setups.  We leave that for extensions.  Do you confirm using the extension fixes the issue?

That being said, we could maybe improve the error message by suggesting to use the `flycheck-haskell` extension in this case.

**Comment by andersk** [comment]
I tried it with `flycheck-haskell` and it works…mostly*.

(Yeah, it was sort of unclear to me what `flycheck-haskell` does that core `flycheck` doesn’t. I guess this is part of it.)

\* It still fails in the same way on the first Haskell file I open, until I reopen it. However, I suspect that’s somehow a result of loading `flycheck` and `flycheck-haskell` from a byte-compiled init file with `use-package`, so I’ll investigate that more before further complaining, unless you happen to know what I did wrong here:

```elisp
(use-package flycheck-haskell
  :ensure t
  :commands flycheck-haskell-setup)

(use-package flycheck
  :ensure t
  :defer 0
  :commands global-flycheck-mode
  :init
  (add-hook 'flycheck-mode-hook #'flycheck-haskell-setup)
  :config
  (global-flycheck-mode))
```

**Comment by fmdkdd** [comment]
It can be tricky to find the right order of loading with `use-package`.  Here is how I set it up for the `flycheck-rust` extension (which has the same role as the haskell one):

https://github.com/fmdkdd/dotfiles/blob/6b4bfa2f52e249c4c1cf5de16efa97c0d1095de6/emacs/.emacs#L251-L267

**Comment by cpitclaudel** [comment]
This part looks dubious:

```
(add-hook 'flycheck-mode-hook #'flycheck-haskell-setup)
```

`flycheck-haskell-setup` just does `(add-hook 'hack-local-variables-hook #'flycheck-haskell-configure)`. Since Flycheck gets enabled after local variables are read, `flycheck-mode-hook` runs after `hack-local-variables-hook`. Thus, in your setup, the `flycheck-haskell-configure` function doesn't get called when ou open your first Haskell file (but it does get called the next time you open a Haskell file, since at that point the hook is properly set up).

The fix would be to call `flycheck-haskell-setup` directly in the `:config` of `flycheck-haskell`'s `use-package` form, I think (but I don't know much about use-package).

**Comment by fmdkdd** [comment]
> This part looks dubious:
> (add-hook 'flycheck-mode-hook #'flycheck-haskell-setup)

That's the documented way (`;;;; Setup` section) to setup `flycheck-haskell`.  If that is not right, then we should probably update the doc :)

cc @sergv

**Comment by andersk** [comment]
@cpitclaudel `use-package` does lazy loading and `:config` only runs after the package is loaded, so that would result in `flycheck-haskell` not being loaded automatically at all.

It seems like it would be ideal if `flycheck-haskell-setup` also ran `flycheck-haskell-configure` immediately if…something? What’s the closest we can get to writing “if it’s not about to also run via `hack-local-variables-hook`”? Ideally we’d also catch the case where the user isn’t using `global-flycheck-mode` and instead manually turns on `flycheck-mode` in a Haskell buffer some time after it’s opened.

**Comment by cpitclaudel** [comment]
> @cpitclaudel use-package does lazy loading and :config only runs after the package is loaded, so that would result in flycheck-haskell not being loaded automatically at all.

Of course, sorry. Would it work to just call `(flycheck-haskell-setup)` in `flycheck`'s `:config` or `:init`, then?
It doesn't handle the more complex case that you outline below.

> It seems like it would be ideal if flycheck-haskell-setup also ran flycheck-haskell-configure immediately if…something? What’s the closest we can get to writing “if it’s not about to also run via hack-local-variables-hook”? Ideally we’d also catch the case where the user isn’t using global-flycheck-mode and instead manually turns on flycheck-mode in a Haskell buffer some time after it’s opened.

I think it's easier than that: we want to write "if it hasn't already run via hack-local-variables-hook".  For that, it should be enough to keep a variable tracking whether we've run that function before. Something like the following, added to `flycheck-haskell-setup`:

```
(unless flycheck-haskell-configured
  (setq-local flycheck-haskell-configured t)
  (flycheck-haskell-configure))
```

But even that is still too complex. A much simpler approach is to make the hook buffer-local, and to make `flycheck-haskell-setup` call `flycheck-haskell-configure` unconditionally.

Then the setup you have should work perfectly. The first time the file is open the configuration happens, and the hook is set up to refresh the configuration if local variables are changed.

This is a breaking change, though, so it might be better to just create a minor mode `flycheck-haskell-mode` that does the above (call `flycheck-haskell-configure` and register the hook when enabled), which we'd recommend in the docs.  Unless we know that most people just add flycheck-haskell-setup to their flycheck-mode-hook?

**Comment by sergv** [comment]
Sorry to catch up a bit late in the process. I think that many people just add `flycheck-haskell-setup` to their `flycheck-mode-hook` (github search yields around 300 matches), which is not surprising given that it was the recommended setup for a long time. I can think of two reasons to keep configuration interface the way it is and not introduce `flycheck-haskell-mode`:
- Backward compatibility
- `flycheck-rust` is configured similarly and seems to get away with it. I think it's somewhat nice for auxiliary flycheck plugins to have similar setup instructions

Therefore I think that `flycheck-haskell-setup` should be fixed to do something sensible for the first time the buffer is opened - it should make an initial configuration and install a hook. If there are no objections I can prepare a PR.

**Comment by cpitclaudel** [comment]
Please do

**Comment by cpitclaudel** [comment]
Thanks!

**Comment by sergv** [comment]
I have fixed the `flycheck-haskell` project. Please try it out and check whether it works for you.

**Comment by andersk** [comment]
Yeah it seems to be working. Thanks!

**Comment by sergv** [comment]
I think we can close the issue for now. Please reopen if anything goes wrong.
